### PR TITLE
Geomap graph (in progress)

### PR DESCRIPTION
**Pull request by nxanil**
@ronakmshah 
The first stage of geo map graph with the basic feature like cluster markers with info window has been implemented. Please have a look.

geomap

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -0,0 +1,7 @@
+REACT_APP_ELASTICSEARCH_HOST=http://localhost:9200
```
Comment (bmukheja) [inline .env.example]: Is there a better way to provide these variables than to store them in a local file?

Comment (nxanil) [reply .env.example]: @bmukheja This is the standard we normally follow.
But we can provide these values through various options like 
1) URL params
2) Through API returning all the settings.
3) Via settings table in DB but currently, we don't have any database.

**Comment by nxanil** [comment]
@ronakmshah @bmukheja 

Added event listener on marker's click. please have a look.

File: .env.example (hunk 1 above)
Comment (bmukheja) [reply .env.example]: I think URL params should be OK for all, as we are already following that. 
@ronakmshah - But for Google api key, we need to make sure that the key is coming from the customer in production and not the same key as dev environment. DO you think VSD guys can help us in that?

**Comment by bmukheja** [comment]
Nice work anil. We need to further discuss some other basic things such as -
1. No street view.
2. Offline availability.
3. Changing the icon for clusterer

**Comment by bmukheja** [comment]
Hi Anil,

Adding a simpler feature into GeoMap. There is a search functionality in the react library we are using, we'd like that to be present on the top of map, being able to search for NSG names. Please check if that can be implemented.

**Comment by bmukheja** [comment]
5. Make the default location as the average of the latitudes and longitudes of all the NSGs

**Comment by bmukheja** [comment]
6. Make the cluster show label if all the NSGs in the cluster have the exact same location. The label would show the same information as the other label on single NSGs.
7. Log the response of the query on console like elasticsearch queries.
8. 3 tiers for clusters - 2-10, 10-50,50+. Each has different size
9. Option to configure icon for cluster(or configure one time internally)

**Comment by nxanil** [comment]
@bmukheja 
Added event listener on marker click. Please have a look

**Comment by bmukheja** [comment]
@nxanil @ronakmshah - Barring the merge issue we discussed on the call, I believe now is the time to talk about the clustering issue on same location. @nxanil Do you have any suggestions? Just as an example, I saw what https://leaflet.github.io/Leaflet.markercluster/example/marker-clustering-realworld.10000.html have done and think they have handled it pretty nicely. If this is something which can done that would be really cool, or we could come up with our own unique way as well

**Comment by nxanil** [comment]
@bmukheja Could you please check it now and get back to me if you have any issues. Click listener seems to be working fine. Just reviewed it.

**Comment by bmukheja** [comment]
@nxanil - Listener works fine, but the following things need to be considered -

0. Add the config file which is still missing
1. I couldn't see links after running it with elasticsearch query, neither are links visible with the test data
2. I have manually added the mouseOver function, but please add mouseOut too. They should toggle infowindow
3. Zoom In & Zoom out animations from leaflet libraryClusterer library.

The comments earlier this are completed. Thanks for your input on this.

**Comment by bmukheja** [comment]
@nxanil - 

4. Geomap is showing "No Data to display" when ES query fails. The ES query is "data1" query, it should work if the primary query "data" is functioning. 
5. Is it possible to have an analysis of system resources this is going to consume, and list out the resource requirements such as System processing power, memory, B/W consumption?
6. Add a marker resolver, a dictionary which resolves a particular field, say "gatewayType" to marker icon. This can be specified by the user/creator of viz. It could default to the default icon in case an icon file is not present.

**Comment by nxanil** [comment]
@ronakmshah @bmukheja 

- added a new feature to scattered the markers present at the same location mentioned in issue #654. See in below screenshots:-

geomap-scatter

geomap-scatter 2

- Also, removed sea routes, satellite views mentioned in issue #663

- Fixed issue #670, configured zoom out limit from default configuration, now you may override it.

- Fixed issue #667.

**Comment by nxanil** [comment]
@ronakmshah 

- Removed landmark markers (hospitals and schools and mountain peaks) 

- Added street labels

- Restrict zoom out limit. Zoom will depend upon the size of the container, therefore, you can override it from configuration as per the visualization, where you were using it.
You can set max/min zoom limit via following variables - 
 `  maxZoom: 18`
  `minZoom: 3`

**Comment by nxanil** [comment]
@ronakmshah 

Fixed issues mentioned in email related to duplicate markers on searching etc. Please have a look and let me know.

**Comment by nxanil** [comment]
@ronakmshah 

This PR has been updated with master and is now using the "vis-graphs" repo.
Therefore, to check this PR,  firstly update the submodule and then change the branch of "vis-graphs" to "feature/geomap" which is placed inside src/lib/vis-graphs.

Once you verified the changes, then kindly merge both the PR i.e. this one, along with "vis-graph's PR https://github.com/nuagenetworks/vis-graphs/pull/2"

Thanks

**Comment by bmukheja** [comment]
Closing this, as all changes from this have been implemented elsewhere

**Pull request closed without merging**